### PR TITLE
UTF-8 parsing

### PR DESCRIPTION
**Pull request by ruancomelli**
Very much a work in progress PR, creating this for CI + coderabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced a robust UTF-8 parsing and validation framework to enhance character encoding handling.
	- Added a command-line option parsing library for improved management of arguments.

- **Build System Changes**
	- Expanded build targets in the Makefile to include new UTF-8 related tests.
	- Streamlined the handling of compiler flags for a more efficient build process.
	- Enhanced documentation generation process through Doxygen integration.

- **Chores**
	- Updated `.gitignore` to exclude specific files and directories from version control.
	- Added a new `.clang-tidy` configuration for enhanced static analysis.
	- Integrated a Git submodule for improved documentation styling.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is a benchmark review for experiment `multiple_mermaid_diagram_types`.
Run ID: `multiple_mermaid_diagram_types/benchmark_2024-09-26T14-38-21_v1-23-0-20-g8a080be57-dirty`.

This pull request was cloned from `https://github.com/imaami/libcanth/pull/2`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
    # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews: []

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request introduces a robust UTF-8 parsing and validation framework, along with command-line option parsing capabilities. It also includes updates to the build system and project configuration files.

### Sequence Diagrams

#### UTF-8 Parsing Sequence Diagram

```mermaid
sequenceDiagram
    participant Client
    participant utf8_parse_next_code_point
    participant utf8_set_state
    participant utf8_push_to_cache

    Client->>utf8_parse_next_code_point: Parse next code point
    utf8_parse_next_code_point->>utf8_set_state: Update state
    utf8_set_state->>utf8_push_to_cache: Push byte to cache
    utf8_push_to_cache-->>utf8_set_state: Return
    utf8_set_state-->>utf8_parse_next_code_point: Return updated state
    utf8_parse_next_code_point-->>Client: Return next byte pointer
```

#### Command-line Option Parsing Class Diagram

```mermaid
classDiagram
    class letopt_state {
        +char** v
        +int c
        +int e
        +char const** q
        +int i
        +int n
        +char* p
    }
    class letopt {
        +letopt_state p
        +bool has
    }
    letopt *-- letopt_state
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Implemented a UTF-8 parsing and validation framework | <ul><li>Created utf8.h header file with UTF-8 parser states and functions</li><li>Implemented utf8.c with core UTF-8 parsing logic</li><li>Added test-utf8.c for testing and demonstrating UTF-8 parsing functionality</li></ul> | `src/utf8.h`<br/>`src/utf8.c`<br/>`src/test-utf8.c` |
| Added a command-line option parsing library | <ul><li>Implemented letopt.h header file with option parsing structures and functions</li><li>Created letopt.c with core option parsing logic</li></ul> | `src/letopt.h`<br/>`src/letopt.c` |
| Updated build system and project configuration | <ul><li>Modified Makefile to include new UTF-8 related tests</li><li>Added a new Makefile in the doc directory for documentation generation</li><li>Updated .gitignore to exclude specific files and directories</li><li>Added .clang-tidy configuration for enhanced static analysis</li></ul> | `src/Makefile`<br/>`doc/Makefile`<br/>`.gitignore`<br/>`.clang-tidy` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/utf8.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        F(15,  ini,    0,    0,   0, 0 ,0) /* initial state, not used as a flag in the lookup table */
+
+/**
+ * @brief UTF-8 parser state enumeration.
+ */
+fixed_enum(utf8_st8, uint8_t) {
+	#define F(n,m,...) utf8_##m = n,
+	UTF8_PARSER_DESCRIPTOR(F)
+	#undef F
+};
```
Comment (SourceryAI) [inline src/utf8.h:78]: **suggestion:** Consider adding comments explaining the purpose of each field

While the field names are somewhat self-explanatory, adding brief comments for each field in the utf8 struct would improve code readability and maintainability, especially for developers who might not be familiar with the intricacies of UTF-8 parsing.

File: src/letopt.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		.v = argv,
+		.c = argc
+	};
+
+	if (state.c > 1 &&
+	    !(state.q = calloc((size_t)state.c, sizeof *state.q)))
+		state.e = errno ? errno : ENOMEM;
+
+	return state;
+}
```
Comment (SourceryAI) [inline src/letopt.c:36]: **suggestion:** Consider using strtoll consistently instead of _Generic

The use of _Generic to choose between strtol and strtoll based on the type of 'n' might be unnecessary. Since you're using int64_t for the result, it would be simpler and more consistent to always use strtoll.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider adding comprehensive unit tests for the UTF-8 parser to ensure correctness across various edge cases.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/d9bf9cdb-ddd8-4a50-9eec-6ca5e29d6540?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>
